File: daily-coding/src/dailyprogrammer/intermediate/_238/FalloutHacking.py
from typing import *
import random

# Hard-coded settings
SETTING_GUESSES = 4
SETTING_FNAME = "popular_words.txt"


class Model:  # Represents an instance of hacking game
    """Instance variable descriptions below
    1) word_bank: all words in txt file
    2) valid_words: valid length words from word bank
    3) game_words: randomly chosen words from valid words
    """
    word_bank = open(SETTING_FNAME, "r").read().splitlines()

    def __init__(self, guesses: int, difficulty: int):
        self.guesses = guesses
        self.difficulty = difficulty
        self.word_length = 4 + difficulty * 2
        self.word_amt = 6 + difficulty * 2
        self.valid_words = self.filter_words()
        self.game_words = self.pick_words()
        self.answer = self.pick_answer()

    def filter_words(self) -> List[str]:
        """Returns words of correct word length from word bank
        Time complexity O(n)
        """
        filtered_words = list()
        for word in Model.word_bank:
            if len(word) == self.word_length:
                filtered_words.append(word)
        return filtered_words

    def pick_words(self) -> List[str]:
        """Returns randomly picked game words from valid words
        """
        game_words = list()
        for i in range(0, self.word_amt):
            game_words.append(
                self.valid_words[random.randint(0, len(self.valid_words))])
        return game_words

    def pick_answer(self) -> str:
        """Returns randomly picked answer word from game words
        """
        return self.game_words[random.randint(0, len(self.game_words))]

    def roll_new(self):
        """Randomly set new answer and game words
        """
        self.game_words = self.pick_words()
        self.answer = self.pick_answer()


class View:
    def __init__(self, model):
        self.model = model

    @staticmethod
    def ask_difficulty() -> int:
        return int(input('Difficulty (0-4)? '))

    def print_words(self):
        for word in self.model.game_words:
            print(word.upper())

    def ask_guess(self) -> str:
        guess = input('Guess ({} left)? '.format(self.model.guesses))
        if guess not in self.model.game_words:
            raise Exception('ask_guess(): Guess word not found in game words')
        return guess


class Controller:  # Managing instances of model and view
    def __init__(self, model, view):
        self.model = model
        self.view = view

    def start_game(self):
        """Begins game loop of guessing
        """
        self.view.print_words()
        for i in range(0, self.model.guesses):
            guess = self.view.ask_guess()
            correct_amt = self.calc_correct(guess)
            print('{}/{} correct'.format(correct_amt, self.model.guesses))
            if correct_amt == self.model.word_length:
                print('You win!')
                break

    def calc_correct(self, guess):
        correct_amt = 0
        for i in range(0, len(guess)):
            if guess[0] == self.model.answer[0]:
                correct_amt += 1
        return correct_amt


def main():
    model = Model(SETTING_GUESSES, View.ask_difficulty())
    view = View(model)
    controller = Controller(model, view)
    controller.start_game()


if __name__ == '__main__':
    main()

File: daily-coding/src/dailyprogrammer/intermediate/_238/FalloutHacking.java
package dailyprogrammer.intermediate._238;

import java.io.File;
import java.io.FileNotFoundException;
import java.net.URL;
import java.util.ArrayList;
import java.util.Scanner;
import java.util.Random;

public class FalloutHacking {
    static final String WORD_FNAME = "popular_words.txt";
    static final int[] WORD_LEN = {4, 6, 8, 10, 12};
    static final int[] WORD_AMT = {6, 8, 10, 12, 14};
    static final int GUESSES = 4;

    public static int inputDifficulty() {
        Scanner scan = new Scanner(System.in);
        System.out.print("Enter difficulty level(0-4): ");
        return scan.nextInt();
    }

    public static ArrayList<String> getWords(int difficulty)
            throws FileNotFoundException {
        /* Scans through entire word text file and puts valid words in list
        Time complexity: O(n)
            n words in word text file
         */
        URL url = FalloutHacking.class.getResource(WORD_FNAME);
        File wordFile = new File(url.getPath());
        Scanner scan = new Scanner(wordFile);
        ArrayList<String> wordList = new ArrayList<String>();

        while (scan.hasNext()) {
            String word = scan.next();
            if (word.length() == WORD_LEN[difficulty])
                wordList.add(word);
        }

        return wordList;
    }

    public static ArrayList<String> pickWords(
            ArrayList<String> wordList, int numWords) {
        /* Randomly picks words from valid words list
         */
        Random rand = new Random();
        ArrayList<String> gameWords = new ArrayList<String>();

        for (int i = 0; i < numWords; i++) {
            int randInt = rand.nextInt(wordList.size());
            gameWords.add(wordList.get(randInt));
        }
        return gameWords;
    }

    public static int findMatchAmt(String guess, String answer) {
        /* Returns number of matching characters between guess and answer
        Assumes guess and answer are same length
         */
        int matchAmt = 0;
        for (int i = 0; i < guess.length(); i++) {
            if (guess.charAt(i) == answer.charAt(i))
                matchAmt++;
        }
        return matchAmt;
    }

    public static void startGame(ArrayList<String> gameWords) {
        /* The "game master"
         */
        Scanner scan = new Scanner(System.in);
        Random rand = new Random();
        // Print out all game words
        for (String gameWord : gameWords)
            System.out.println(gameWord.toUpperCase());
        // Randomly pick an answer word
        String answer = gameWords.get(rand.nextInt(gameWords.size()));
        // Ask for a guess, then print number letters correct
        for (int i = GUESSES; i >= 0; i--) {
            System.out.printf("\nGuess (%d left)? ", i);
            String guess = scan.next();
            int matchAmt = findMatchAmt(guess, answer);
            System.out.printf("%d/%d correct", matchAmt, guess.length());
            if (matchAmt == guess.length()) {
                System.out.println("\nYou win!");
                break;
            }
        }
    }

    public static void main(String[] args) throws FileNotFoundException {
        int diff = inputDifficulty();
        ArrayList<String> wordList = getWords(diff);
        ArrayList<String> gameWords=  pickWords(wordList, WORD_AMT[diff]);
        startGame(gameWords);
    }
}
